MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressMediaSanitize.c

```diff
@@ -0,0 +1,473 @@
+                                 SectorOwBuffer // overwrite buffer
+                                 );
+    }
+
+    //
+    // Reset SectorOffset back to zero if another pass on namespace is needed
+    //
+    SectorOffset = 0;
+  }
+
+  gBS->RestoreTPL (OldTpl);
+
+  return Status;
+}
+
+/**
+  Purge Media utilizes transport native Sanitize operations. Sanitize specific
+  purge actions include: overwrite, block erase, or crypto erase.
+
+  Functions are defined to erase and purge data at a block level from mass
+  storage devices as well as to manage such devices in the EFI boot services
+  environment. Sanitization refers to a process that renders access to target
+  data on the media infeasible for a given level of effort.
+
+  NOTE: This operation is a blocking call.
+
+  @param  This             Indicates a pointer to the calling context.
+  @param  MediaId          The media ID that the write request is for.
+  @param  PurgeAction      The purage action (overwrite, crypto erase, block erase).
+  @param  OverwritePattern 32-bit pattern to overwrite on media (for overwrite).
+
+  @retval EFI_SUCCESS           The data was written correctly to the device.
+  @retval EFI_WRITE_PROTECTED   The device can not be written to.
+  @retval EFI_DEVICE_ERROR      The device reported an error while performing the write.
+  @retval EFI_NO_MEDIA          There is no media in the device.
+  @retval EFI_MEDIA_CHNAGED     The MediaId does not matched the current device.
+  @retval EFI_BAD_BUFFER_SIZE   The Buffer was not a multiple of the block size of the device.
+  @retval EFI_INVALID_PARAMETER The write request contains LBAs that are not valid,
+                                or the buffer is not on proper alignment.
+
+**/
+EFI_STATUS
+EFIAPI
+NvmExpressMediaPurge (
+  IN EFI_MEDIA_SANITIZE_PROTOCOL  *This,
+  IN UINT32                       MediaId,
+  IN UINT32                       PurgeAction,
+  IN UINT32                       OverwritePattern
+  )
+{
+  NVME_DEVICE_PRIVATE_DATA  *Device;
+  EFI_BLOCK_IO_MEDIA        *Media;
+  UINT32                    SanitizeAction;
+  UINT32                    NoDeallocate;
+  UINT32                    NamespaceId;
+  EFI_TPL                   OldTpl;
+  EFI_STATUS                Status;
+
+  //
+  // Check parameters.
+  //
+  if (This == NULL) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  Device      = NVME_DEVICE_PRIVATE_DATA_FROM_MEDIA_SANITIZE (This);
+  NamespaceId = Device->NamespaceId;
+  Media       = &Device->Media;
+
+  if ((MediaId != Media->MediaId) || (!Media->MediaPresent)) {
+    return EFI_MEDIA_CHANGED;
+  }
+
+  if (PurgeAction & EfiBlockSanitizeOverwrite) {
+    SanitizeAction = SANITIZE_ACTION_OVERWRITE;
+  } else if (PurgeAction & EfiBlockSanitizeBlockErase) {
+    SanitizeAction = SANITIZE_ACTION_BLOCK_ERASE;
+  } else if (PurgeAction & EfiBlockSanitizeCryptoErase) {
+    SanitizeAction = SANITIZE_ACTION_CRYPTO_ERASE;
+  } else {
+    SanitizeAction = 0;

```

Comment:
Yes, it does. The call to NvmExpressSanitize() will perform other checks internally at the NVMe layer... so I was trying to avoid doing too much NVMe-specific checking in this MediaPurge function. But I do see the value of adding more "official" checking in this function to demonstrate what is and isn't acceptable at this protocol layer. I will update this mapping to include checks for invalid and unsupported values.

A purge action of 0 means perform no sanitize action (i.e a Sanitize no-op). My idea was to let the end storage device fail the transport-specific "Sanac" invalid field. I want to stay away from the transport-specific type checking that should be performed by the end device. Do you see a case to where I should put the check here instead?